Sofa/Component/StateContainer/src/sofa/component/statecontainer/init.cpp

```diff
@@ -55,6 +60,39 @@ void init()
     }
 }
 
+void registerObjects(sofa::core::ObjectFactory* factory)
+{
+    static bool registered = false;
+
+    if (!registered)
+    {
+        using namespace sofa::defaulttype;
+        
+        // Registration with explicit commit
+        // MappedObject
+        core::RegisterObject("Mapped state vectors")
+            .add< MappedObject<Vec1Types> >()
+            .add< MappedObject<Vec3Types> >(true) // default template
+            .add< MappedObject<Vec2Types> >()
+            .add< MappedObject<Vec6Types> >()
+            .add< MappedObject<Rigid3Types> >()
+            .add< MappedObject<Rigid2Types> >()
+            .commit(factory);
+
+        // Registration with RAII-style commit
+        // MechanicalObject
+        core::RegisterObject("mechanical state vectors", factory)
+            .add< MechanicalObject<Vec3Types> >(true) // default template
+            .add< MechanicalObject<Vec2Types> >()
+            .add< MechanicalObject<Vec1Types> >()
+            .add< MechanicalObject<Vec6Types> >()
+            .add< MechanicalObject<Rigid3Types> >()
+            .add< MechanicalObject<Rigid2Types> >();
```

Comment:
I am thinking about the best design to register objects in the factory.

I think your proposition of code is error prone. You can forget to commit, or to add the factory as the second argument. It's probably due to the old design, and maybe we can think of a better design. Ideally, I would rather write something more explicit such as:

```cpp
const auto objectToRegister = core::RegisterObject(...).add<...>().add<...>();
factory->registerObject(objectToRegister);
```

And why not going further:

```cpp
factory->registerObject(core::RegisterObject("Mechanical state vectors in 2d").add<MechanicalObject<Vec2Types>>;
factory->registerObject(core::RegisterObject("Mechanical state vectors in 3d").add<MechanicalObject<Vec3Types>>;
...
```

Even the class `RegisterObject` bothers me. Nothing prevents not to call `add`. Maybe something like:

```cpp
template<class T>
class ObjectToRegister
{...};

factory->registerObject("Mechanical state vectors in 2d", ObjectToRegister<MechanicalObject<Vec2Types>>>;
```

or

```cpp
factory->registerObject<MechanicalObject<Vec2Types>>("Mechanical state vectors in 2d");
factory->registerObject<MechanicalObject<Vec3Types>>("Mechanical state vectors in 3d", ObjectFactory::defaultTemplate);

//ObjectFactory::defaultTemplate would be a strongly-typed boolean.
```

This way, you are obliged to use the variable `factory`, and no more a `RegisterObject` object which can be misleading, as it does not register object per se.

And if you want to keep the same description (you can always use a variable), you can write a parameter pack which would look like:

```cpp

class ObjectFactory
{
template<class ComponentType>
void registerObject();

template<class ComponentType, class RestOfComponentTypes...>
void registerObject()
{
    this->registerObject<ComponentType>();
    this->registerObject<RestOfComponentTypes...>();
}
};


factory->registerObject<
    MechanicalObject<Vec1Types>,
    MechanicalObject<Vec2Types>,
    DefaultTemplate<MechanicalObject<Vec3Types>>
>("Mechanical state vectors");
```

In summary, I would not continue to use a `RegisterObject` class. Note that I was thinking about interface only, not about the implementation which can be problematic, neither the compatibility with the old design.